src/util/shared_exclusive_lock.c

```diff
@@ -12,98 +12,146 @@
+SharedExclusiveLockType SharedExclusiveLock_Acquire(RedisModuleCtx *ctx) {
+  // First, acquire the alternative lock. Only one thread can try to acquire either the GIL or the alternative lock at a time.
+  // - we don't hold the internal lock (it's never held out of shared_exclusive_lock.c)
+  // - we may hold the GIL (allowed)
+  pthread_mutex_lock(&GILAlternativeLock);
   while (true) {
-    int rc = REDISMODULE_ERR;
-    // GILAlternativeLockHeld condition check is needed to avoid race condition with the Release.
-    if (GILOwned && !GILAlternativeLockHeld) {
-      rc = pthread_mutex_trylock(&GILAlternativeLock);
-      if (rc == 0) {
-        // We acquired the alternative lock, we can return.
-        GILAlternativeLockHeld = true;
-        pthread_mutex_unlock(&InternalLock);
-        return Internal_Locked;
-      }
-      rc = REDISMODULE_ERR;
-    } else {
-      if (!acquireInternalLock) {
-        rc = RedisModule_ThreadSafeContextTryLock(ctx);
-      }
+    SharedExclusiveLockType lockType = Unlocked;
+    // Attempt to acquire the GIL, according to the internal state.
+    pthread_mutex_lock(&InternalLock);
+    if (GIL_lent) {
+      // GIL is lent by the main thread, mark that we borrow it.
+      GIL_borrowed = true;
+      lockType = Borrowed;
+    } else if (RedisModule_ThreadSafeContextTryLock(ctx) == REDISMODULE_OK) {
+      // GIL is not lent by the main thread, but we managed to acquire it.
+      // Note: Under assumption (2), either the GIL is lent, or we don't hold it, so we can attempt to acquire it.
+      lockType = Owned;
     }
-    if (rc == REDISMODULE_OK) {
-      RS_LOG_ASSERT(!GILAlternativeLockHeld, "If we acquired the GIL, the alternative lock should not be held by another thread.");
-      pthread_mutex_unlock(&InternalLock);
-      return GIL_Locked;
+    pthread_mutex_unlock(&InternalLock);
+
+    if (lockType != Unlocked) {
+      return lockType; // We have exclusive access
     }
 
+    // Couldn't acquire the GIL, wait (for a short time or until signaled) before trying again.
     struct timespec timeout;
-    clock_gettime(CLOCK_REALTIME, &timeout);
-    timeout.tv_nsec += TIMEOUT_NANOSECONDS;
-    // Handle nanosecond overflow to comply with POSIX timespec requirements
-    if (timeout.tv_nsec >= 1000000000) {
-      timeout.tv_sec += timeout.tv_nsec / 1000000000;
-      timeout.tv_nsec %= 1000000000;
-    }
-    pthread_cond_timedwait(&TryLockCondition, &InternalLock, &timeout);
+    set_timeout(&timeout);
+    pthread_cond_timedwait(&GILAvailable, &GILAlternativeLock, &timeout);
   }
 }
 
+// Assumptions:
+// 1. The caller has previously acquired the lock by calling SharedExclusiveLock_Acquire
+// 2. The value of 'type' is the result of the previous call to SharedExclusiveLock_Acquire
 void SharedExclusiveLock_Release(RedisModuleCtx *ctx, SharedExclusiveLockType type) {
-  if (type == Internal_Locked) {
+  if (type == Borrowed) {
```

Comment:
Even though it's useful to get the lock type from `SharedExclusiveLock_Acquire` for testing or logging, I think the Release function should make its decision based on the `GIL_borrowed` variable

Reply:
I prefer to rely as little as possible on globals, and testing is not at all a minor issue